linode/manager
# feat: [M3-6731] – Add VPC and Firewall sections in Linode Create flow

## Description 📝
- Add VPC section in Linode Create flow
- Add Firewall section in Linode Create flow

## Major Changes 🔄
- `VPCs` added to region capabilities
- `linodeInterfaceSchema` for individual interfaces, `linodeInterfacesSchema` for the actual `interfaces` property in Linode Create payload
- `refetchOnWindowFocus` added to `useVPCsQuery`
- New `VPCPanel` & `SelectFirewallPanel` components that are present in Linode Create flow
  - New `linodeCreateWithFirewall` feature flag that determines whether `SelectFirewallPanel` is displayed or not in Linode Create flow
- Conditional copy for "Private IP" in "Add-ons" panel in Linode Create flow

## Preview 📷
<details>
<summary>VPC and Firewall panels</summary>

![Screenshot 2023-09-21 at 2 10 29 PM](https://github.com/linode/manager/assets/114682940/2d028fa2-8f33-4407-95fc-7656c7ff2af7)
</details>

<details>
<summary>"Private IP" conditional copy</summary>

![Screenshot 2023-09-21 at 2 11 36 PM](https://github.com/linode/manager/assets/114682940/4faf236a-a363-4d65-ae39-9edc6405fcfa)
</details>

<details>
<summary>"VPC Assigned" in summary paper</summary>

![Screenshot 2023-09-21 at 2 23 18 PM](https://github.com/linode/manager/assets/114682940/08292d7e-7f06-44af-8978-9001e59b235b)

</details>

## How to test 🧪
With a properly-tagged account or with the VPC feature flag on, 

1. Confirm the presence of the VPC & Firewall panels in the Linode Create flow
2. For the VPC panel: 
a) Nothing below the "Create VPC" link shows when the VPC dropdown is set to "None"
b) the VPC dropdown options are only those from the same region as the region selected for the Linode
c) the Subnet dropdown options are only those subnets belonging to the selected VPC
d) the "Auto-assign a VPC IPv4 address for this Linode in the VPC" checkbox is checked by default
e) unchecking the "Auto-assign a VPC IPv4 address for this Linode in the VPC" checkbox reveals a "VPC IPv4" text field which must be filled out (attempting to create the linode without this should yield a field error)
f) expectations for the payload based on the checkboxes: a checked "Auto-assign a VPC IPv4 address for this Linode in the VPC" checkbox should result in `ipv4.vpc` not being present in the payload, while an unchecked one should result in the contents of the "VPC IPv4" text field populating `ipv4.vpc` in the payload; a checked "Assign a public IPv4 address for this Linode" should result in `ipv4.nat_1_1` being set to `any`, while an unchecked one should result in `ipv4.nat_1_1` not being present in the payload the original tab: the VPC dropdown should have your newly-created VPC included
g) Clicking the "Create VPC" link and completing that flow in the new tab, then returning to the Linode Create flow in
3. For the Firewall panel:
a) Your firewalls should be listed as options
b) You can add `linodeCreateWithFirewall: false` somewhere between L30-L33 in `withFeatureFlagConsumer.container.ts` to confirm the Firewall panel is not visible without the feature flag being on
4. The Summary paper at the bottom should say "VPC Assigned" if you've assigned a VPC
5. Creating a non-VPC, non-firewalled linode with and without a VLAN
6. Creating a linode with both a VPC and a VLAN

With a non-tagged account & the VPC feature flag off, confirm (1) above is not present, and confirm normal Linode creation functionality works as expected.
